Full Localization

ThePoultryMan opened the pull request:
## Summary

Adds localization to all text instances, making the app fully translatable.

## Specifics

* Almost all translatable text keys are formatted in SCREAMING_SNAKE_CASE, and begin with the prefix `TR_` making it clear whether something is translatable to not.
* Translation files are named based off of country and locale, i.e. `en_us.csv`.
* Translation files are sorted alphabetically, except for `key` which is left at the top.
* Language can be selected manually by changing the language setting in the Settings > Remix Settings
* ~~Language selection is saved in `preferences.cfg`~~

## Other Changes

* `FormattedLocalizedLabel` was introduced, allowing for localizations to be intermixed with more hardcoded data.
  * ~~Currently only used by the window size label.~~
* ~~All executable-based save locations have been replaced with `user://` based save locations. This guarantees that files can be written without issue. See [File Paths in Godot projects](https://docs.godotengine.org/en/latest/tutorials/io/data_paths.html) for more information.~~
* ~~A new config file, `preferences.cfg`, is used to handle saving language selection.~~
* A utility class, `Util.gd`, has been introduced. ~~Currently, it only language to locale-code conversions.~~

## To-do

- [x] Specify en_us.csv
  - Godot seems to autodetect English as `en`, without any locale, so I've made the "en_us" the overarching English translation.
- [x] Finish localizing
- [x] Fix `FormattedLocalizedLabel` issues
- [x] Implement the ability to choose language

ThePoultryMan committed:
refactor: property menu translations

ThePoultryMan committed:
refactor: translation prefix change + left editor translation

ThePoultryMan committed:
refactor: movement tab translation

ThePoultryMan committed:
refactor: add localization to wiggle tab

ThePoultryMan committed:
refactor: animation tab localization

ThePoultryMan committed:
refactor: model tab localization

ThePoultryMan committed:
refactor: light settings localization

ThePoultryMan committed:
refactor: tests localization

ThePoultryMan committed:
refactor: top bar localization

ThePoultryMan committed:
fix: use proper save paths

ThePoultryMan committed:
refactor: specify locale

ThePoultryMan committed:
refactor: miscellaneous localizations

ThePoultryMan committed:
refactor: state remap localization

ThePoultryMan committed:
refactor: name placeholder localization

ThePoultryMan committed:
fix: give credits minimum size

ThePoultryMan committed:
feat: language selection

ThePoultryMan committed:
feat: settings localization

ThePoultryMan committed:
fix: miscellaneous localizations

MudkipWorld commented:
I had been lurking over this PR and I am amazed. I really appreciate your help with making localization easier for future releases :D

ThePoultryMan committed:
refactor: move lang file to subfolder

ThePoultryMan committed:
fix: update file

ThePoultryMan commented:
I believe that all items have been added to "en_us.csv". However, some "non-standard" keys exist, those being tabs.

The only thing left to do is add an auto-transition system from the old config files to the new one, so changes are preserved.

ThePoultryMan committed:
refactor: consolidate preferences into single interface

ThePoultryMan committed:
fix: remove old preferences file

ThePoultryMan committed:
feat: various missed localizations

ThePoultryMan commented:
# Full List of Changes
* All text, that I could find, has been localized.
* File save paths are no longer based on executable location, and instead use the `user://` prefix with the custom path name "pngtuber-remix". Read more [here](https://docs.godotengine.org/en/latest/tutorials/io/data_paths.html).
  * This should fix all issues where files cannot be written due to invalid permissions. For example, trying to write to `/bin` on Linux.
  * Old preference files are migrated to the new system, and the old file is only deleted if the new file is successfully created.
* All preferences have been brought into a single file, "preferences.prdat", and can be accessed via the `Preferences` global class.
* Auto-saves are moved to `user://` space.
* Languages can be selected within settings in the `Remix Settings` tab.
  * Language is set to automatic by default, however, this ignores locale, so US English would be represented by English.
* Some small formatting changes were made to the mini-tutorial's layout in order to make auto-wrapping work.

ThePoultryMan committed:
feat: move auto saves

MudkipWorld commented:
You are doing a great job, but please don't change the preferences, autosaves, etc.. folder to user://
This could easily lead to people losing their files and data. Not everyone knows how to navigate or know that user:// exists. Which makes debugging with users way harder. I understand it makes it easier to load preferences from older versions, but sometimes preferences, lipsync training and stuff breaks between updates. This is why I don't recommend this. Otherwise the localization related improvements are great!
I really hope you understand.

MudkipWorld commented:
Here is a an example fix maybe. By default everything saves in the executable's directory, but if something wrong happens. It would save in the user://
Another fix would be to have the user choose the directory where their preferences, autosaves and such would be loaded/ save and that is would be a configuration file in user:// instead.
If this will be done, there should be a button added somewhere to open the user:// 
I have seen people lose models from Pngtuber plus due to being locked to User:// (which gets erased if new windows is installed, etc..)
This is why I stayed away from using this file path from the first place. 
Please either implement one of the mentioned fix or revert this change. I don't want anyone to lose their models by accident

ThePoultryMan committed:
revert: switch from user:// back to executable location

ThePoultryMan committed:
feat: command line option for specify save data path

ThePoultryMan committed:
fix: crash if lipsync file data doesn't exist

ThePoultryMan commented:
Now, by default, all data is stored next to the executable. However, if the new `--save-data-path=something`  argument is passed on launch, the data will populate the provided folder instead.

Note that the preferences file is now migrated to "preferences-v2.prdat", but it is still located either next to the executable, or wherever the command line option says.

That should allow the app to retain portability, but for anyone that wants/needs a more "installed" approach, that is available.

~~Edit: "~" paths aren't handled on Linux, so almost ready, just have to figure that out.~~
Edit 2: "~" paths aren't handled, but the `HOME` environment variable can be used in its place. So the command line feature is complete.

ThePoultryMan committed:
fix: merge errors

ThePoultryMan commented:
I 100% thought i marked this ready for review. I'll fix merge errors and do some final testing, and then _actually_ mark it ready for review.

MudkipWorld commented:
Yeah, really sorry :'))
i had been doing a lot of fixes in the source code and stuff, if you have any issues with the merge/ merge fix, tell me

ThePoultryMan committed:
fix: preferences saving

ThePoultryMan committed:
fix: training saving

ThePoultryMan committed:
fix: autosave saving

ThePoultryMan committed:
fix: state panel

ThePoultryMan committed:
feat: translation checker

ThePoultryMan committed:
feat: add missing translations

MudkipWorld commented:
what is the estimated time you think you can get this working. i don't want to keep bullying you with unmerged stuff xD

ThePoultryMan committed:
feat: tooltip translation

ThePoultryMan committed:
feat: button translations

ThePoultryMan committed:
feat: exclude translation checker from export

ThePoultryMan committed:
Merge remote-tracking branch 'origin/UIRework' into feat/i18n

ThePoultryMan commented:
Should be ready now, let me know if anything needs changing.

ThePoultryMan commented:
# All Changes

* Settings, autosaves, and lipsync data now all respect the new command line option `save-data-path`.
* Settings now includes an option for language.
* Introduced `translation_checker.gd`, a tool script that, when run, will iterate through all nodes in the selected scene, and report back which ones may contain untranslated text.
* Translation files were moved and renamed.
* All text that I could find was added to the `en_us` translation.

Beyond that, all other statements made in my previous comments should ring true.

MudkipWorld commented:
Will test your PR tomorrow. Thanks you so much for your work!
